# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [4]:
def addTwo(x):
    return x + 2

print(addTwo(4))

#LAMBDA VERSION
print((lambda x: x + 2)(4))
a_num = 4
print((lambda x: x + 2)(a_num))

6
6
6


#### Saving to a Variable

In [5]:
f_test = lambda x: x + 2
print(f_test(4))

6


#### Multiple Inputs

In [7]:
# Multiple inputs with no variable
print((lambda x, y, z: x * y * z)(3, 5, 8))

# Multiple inputs with a variable attached
x_test = lambda x, y, z: x * y * z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [8]:
def multiply(f, num):
    '''
        f expects a lambda function
        num expects a number
    '''
    return f(num)

multiply(lambda x: x * x, 4)

16

#### Returning a Lambda from a Function

In [14]:
def multiply_test(num):
    return num * 4

def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(returnFunc())
print(f_return(4))

def returnLam(b, c):
    return lambda x, a: x + a + b + c
r_lamb = returnLam(4, 5)
print(r_lamb(5, 5))

<function returnFunc.<locals>.multiply at 0x0000022FB5B804A0>
8
19


#### If Statements within Lambdas

In [16]:
f_condition = lambda num : num * 2 if num > 10 else num + 2

print(f_condition(8))
print(f_condition(12))
print(f_condition(10))

10
24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [20]:
f = lambda x, y, z: x * y * z

print(f(3, 3, 3))

27


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [24]:
def squared(num, num2):
    if num < 10 and num2 < 10:
        return num ** 2, num2 ** 2
    else:
        return num, num2
        
numbers = [4, 11, 20, 3, 15, 20]
more_nums = [4, 10, 3, 2, 6]

squared_nums_map = list(map(squared, numbers, more_nums))
print(squared_nums_map)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [34]:
numbers = [1, 5, 14, 8, 12, 9, 4, 3]
numbers_2 = [2, 6, 15, 10, 11, 13, 16, 7]

lamb_number = lambda a, b : a*b if a > b else a+b

used_map = list(map(lamb_number, numbers, numbers_2))
print(used_map)

used_map_2 = list(map((lambda a, b : (a+b) if a > b else (a*b)), numbers, numbers_2))
print(used_map)

[3, 11, 29, 18, 132, 22, 20, 10]
[3, 11, 29, 18, 132, 22, 20, 10]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [38]:
used_map_3 = list(map((lambda n: n*n-1), numbers))
print(used_map_3)

[0, 24, 195, 63, 143, 80, 15, 8]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [42]:
names = ["Bob", "Andy", "Max", "Evan", "Angelica"]

def a_names(name):
    if name[0].lower() == "a":
        return True
    elif name[0].lower() == "b":
        return True
    else:
        return False
new_names_list = list(filter(a_names, names))
print(new_names_list)

['Bob', 'Andy', 'Angelica']


#### Using Lambda's with Filter()

In [50]:
new_names_lambda = list(filter(lambda name: True if name[0].lower() == "a" else False, names))
print(new_names_lambda)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [60]:
import statistics

score = [12, 64, 15, 83, 46, 18, 39, 25, 19, 5]

find_the_mean = list(filter(lambda s: True if s >= statistics.mean(score) else False, score))
print(find_the_mean)

[64, 83, 46, 39]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [65]:
from functools import reduce

list_1 =[2, 4, 6, 8, 10]

def addNums(num1, num2):
    return num1 + num2

result_add = reduce(addNums, list_1)

print(result_add)

def subtractNums(num1, num2):
    return num1 - num2

result_subtract = reduce(subtractNums, list_1)

print(result_subtract)

30
-26


#### Using Lambda's with Reduce()

In [66]:
result_lamb = reduce(lambda x, y: x + y, list_1)
print(result_lamb)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [67]:
my_list = [1, 2, 3, 4]

result = reduce((lambda a, b: a * b), my_list)
print(result)

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [72]:
def addNums(num):
    if num <= 1:
        print("addNums(1) = 1")
        return num
    else:
        print(f"addNums({num}) = {num} + addNum({num - 1})")
        return num + addNums(num -1)
        
addNums(10)

addNums(10) = 10 + addNum(9)
addNums(9) = 9 + addNum(8)
addNums(8) = 8 + addNum(7)
addNums(7) = 7 + addNum(6)
addNums(6) = 6 + addNum(5)
addNums(5) = 5 + addNum(4)
addNums(4) = 4 + addNum(3)
addNums(3) = 3 + addNum(2)
addNums(2) = 2 + addNum(1)
addNums(1) = 1


55

#### Writing a Factorial Function

In [81]:
def factorial(num):
    if num <= 1:
        return num
    else:
        return num * factorial(num - 1)

factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [84]:
def subtractNums(num):
    if num <= 1:
        return num
    else:
        return num - subtractNums(num - 1)

subtractNums(10)

5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [88]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start
        start += step

for i in my_range(20, start = 2):
    my_generator_value = i
    print(my_generator_value)

my_range(20, start = 2)

2
4
6
8
10
12
14
16
18


<generator object my_range at 0x0000022FB5D23C60>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops
# def my_range(stop, start, step = 2):
#     yield start
#     start += step
# DON'T DO THIS!

for i in my_range(20, start = 2):
    my_generator_value = i
    print(my_generator_value)

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [96]:
def squareGenerator(start, stop, step = 1):
    while start > stop:
        yield start
        start -= step

for i in squareGenerator(10, 0):
    gen_value = i*i
    print(gen_value)

100
81
64
49
36
25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [137]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

filtered_places = list(filter(lambda p: False if p.strip() == "" else True, places))

print(filtered_places)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [115]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

def resort(l, n):
    for a in range(2):
        n.append(1)
        for full_name in n:
            if type(full_name) != str:
                break
            new_list = []
            for name in full_name.split(" "):
                new_list.append(name)
            new_list.reverse()
            new_name = ""
            for item in new_list:
                new_name += item + " "
            n.append(new_name.rstrip())

        for i in range(len(n)):
            del n[0]
            if n[0] == 1:
                del n[0]
                break
        if a == 0:
            l(n)
    print(n)

    
resort(lambda x: x.sort(key=str.lower), author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [246]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
converted_places = []

for i in range(len(places)):
    converted = tuple(map((lambda n: (n * 1.8) + 32 if type(n) == int else n), places[i]))
    converted_places.append(converted)
    
print(converted_places)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [155]:
def FibSeq(n, i=0, f=1, s=1): #First int how many times do you want to go up to (required), second int is the step,
# and the third and forth int is the additions for fibonacci sequence
    if i > n:
        return n
    else:
        print(f"Iteration {i}: {f}")
        return i + FibSeq(n, i + 1, s, f + s)

FibSeq(5)

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8


20